[Inductor Cutlass backend] Tolerate dynamic shapes

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121492
* #124577
* #124576
* #124575
* #124574
* #124107
* #121734
* #123932
* #123930
* __->__ #121497

Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

kadeng committed:
[Inductor Cutlass backend] Tolerate dynamic shapes

Previously, when the Cutlass backend was
enabled, using dynamic shapes could lead
to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends,
then an ATen Kernel is used as fallback,
even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was
enabled, using dynamic shapes could lead
to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends,
then an ATen Kernel is used as fallback,
even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was
enabled, using dynamic shapes could lead
to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends,
then an ATen Kernel is used as fallback,
even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was
enabled, using dynamic shapes could lead
to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends,
then an ATen Kernel is used as fallback,
even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano commented on torch/_inductor/kernel/mm.py:
```diff
@@ -136,26 +138,64 @@ def tuned_mm(mat1, mat2, *, layout=None):
+
+
+def _is_static_problem(inputs_tensors, layout):
+    # checks whether all input tensors and the output layout
+    # have a static shape by attempting to convert the dimensions
+    # to int
```
You probably want to use something like the `is_statically_known_list_of_ints` implemented in the `wrapper.py`, which uses the guards information that we already have to simplify the expression.

kadeng replied:
Thanks for the code pointer. I looked it up, and now I'm not sure if that's what I want to use here. I see that the implementation underlying is_statically_known_list_of_ints is this function: https://github.com/pytorch/pytorch/blob/a157a0d00de39fa15f1fd7a5c7c49865bac04f01/torch/_inductor/codegen/wrapper.py#L1527-L1532 which differs from my implementation (just attempt casting to int), by first calling  V.graph._shape_env._maybe_evaluate_static(x) on each dimension and then attempting to cast x to an int. What's surprising to me is that the result of that _maybe_evaluate_static call is just being discarded. I'm not too familiar with the dynamic shape inference system: Does _maybe_evaluate_static have a side effect that would be important here?

lezcano replied:
It doesn't have any side-effect. What it tries to do is use the knowledge from the guards / asserts that the user may have set, etc to evaluate the expression. Sometimes things that look dynamic, may be possibly static after some asserts in the code.

lezcano replied:
Ah, yes, you found a bug there lol

kadeng replied:
I started doubting I understood how this is working ;) I'll prepapre a small fix then...

kadeng replied:
Created https://github.com/pytorch/pytorch/pull/121808

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Rebased on top of origin/viable/strict again on "[Inductor Cutlass backend] Tolerate dynamic shapes"


Previously, when the Cutlass backend was enabled, using dynamic shapes could lead to exceptions during JIT.

With this change, there are guards in place to just disable the Cutlass backend if dynamic dimensions are involved.

In addition, if no choices for a GEMM are available using the selected backends, then an ATen Kernel is used as fallback, even if the ATen backend is not enabled.

Test:
CI
Additional unit test in test_cutlass_backend.py

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng commented:
Just FYI: I will be on PTO soon ( back after easter Monday ), I won't necessarily be able to react to comments until I return.

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng commented on test/inductor/test_aot_inductor.py:
```diff
@@ -2646,7 +2645,6 @@ def fail_non_abi_compatible_cuda(is_skip=False):
     "test_dynamic_cat": fail_minimal_arrayref_interface(),
     # https://github.com/pytorch/pytorch/issues/122978
     "test_dynamic_scalar": fail_stack_allocation(is_skip=True),
-    "test_dynamic_smem_above_default_limit": fail_with_and_without_stack_allocation(),
```
These tests don't fail anymore with this diff, so we need to remove them from the list of expected failures..

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

jansel reviewed (changes requested):
Re-request review once this is ready for me to look at and tests are passing.

kadeng committed:
Update

[ghstack-poisoned]

kadeng committed:
Update

[ghstack-poisoned]

kadeng commented:
The remaining test failures appear unrelated. They don't touch any code that's modified ( e.g. mm.py ), the seabotnet test is obviously flaky when looking at hud.pytorch.org, and I cannot reproduce the failures when I run them locally.

kadeng commented:
@pytorchbot merge --ignore-current

cpuhrsch commented:
Just fyi @kadeng this broke some CUTLASS extensions for quantization, but somehow was still able to land. So we were either missing tests or this stack deleted the tests. This might already be resolved, I'm only adding this comment for reference. cc @HDCharles @alexsamardzic

alexsamardzic commented:
Just posted a fix for CUTLASS-based mixed dtypes GEMM auto-tuning for Ampere: [PR 129058](https://github.com/pytorch/pytorch/pull/129058) - let's continue the discussion on the comments section of this PR.
